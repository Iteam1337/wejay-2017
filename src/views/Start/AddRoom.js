// @flow

import React, { Component } from 'react'
import { graphql } from 'react-apollo'
import { startQuery } from './Start'
import gql from 'graphql-tag'

type Props = {
  mutate: Function,
}

type State = {
  roomName: string,
}

export class AddRoom extends Component<Props, State> {
  state = {
    roomName: '',
  }

  addRoom = (event: SyntheticInputEvent<HTMLFormElement>) => {
    event.preventDefault()

    const { mutate } = this.props
    const { roomName } = this.state

    mutate({
      variables: {
        roomName,
      },
      optimisticResponse: {
        addRoom: {
          name: roomName,
          __typename: 'Room',
        },
      },
      update: (store, { data: { addRoom } }) => {
        const data = store.readQuery({ query: startQuery })

        data.rooms.push(addRoom)

        store.writeQuery({ query: startQuery, data })
      },
    })

    this.setState({
      roomName: '',
    })
  }

  updateRoomName = (event: SyntheticInputEvent<HTMLInputElement>) => {
    this.setState({
      roomName: event.target.value,
    })
  }

  render () {
    const { roomName } = this.state

    return (
      <div className="AddRoom">
        <form method="POST" onSubmit={this.addRoom}>
          <input
            className="Input"
            onChange={this.updateRoomName}
            placeholder="Room name"
            type="text"
            value={this.state.roomName}
          />
          <button
            className="Rooms__button"
            disabled={roomName.length === 0}
            type="submit">
            Add new room
          </button>
        </form>
      </div>
    )
  }
}

const AddRoomMutation = gql`
  mutation addRoom($roomName: String!) {
    addRoom(roomName: $roomName) {
      name
    }
  }
`

export default graphql(AddRoomMutation)(AddRoom)
