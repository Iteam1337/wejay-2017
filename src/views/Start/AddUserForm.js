// @flow

import React from 'react'
import { Formik, Form } from 'formik'
import yup from 'yup'
import Button from '../../components/Button/Button'
import FormInput from '../../components/FormElements/FormInput'

export type FormValues = {
  email: string,
}

type Props = {
  saveUser: (formValues: FormValues) => Promise<void>,
}

const validateUser = yup.object().shape({
  email: yup
    .string()
    .email()
    .required(),
})

const AddUserForm = ({ saveUser }: Props) => {
  return (
    <Formik
      initialValues={{ email: '' }}
      onSubmit={saveUser}
      render={({ isValid }) => (
        <Form>
          <FormInput dataCy="email" name="email" placeholder="E-mail" />

          <Button disabled={!isValid} type="submit">
            Save user
          </Button>
        </Form>
      )}
      validationSchema={validateUser}
    />
  )
}

export default AddUserForm
