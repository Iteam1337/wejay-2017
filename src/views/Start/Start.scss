.Start {
  align-items: center;
  display: flex;
  height: 100vh;
  justify-content: center;
}

.Start__content {
  background: #fff;
  border-radius: 3px;
}

.Start__rooms {
  padding: 40px;
}

.Start__separator {
  background-color: rgba(0, 0, 0, 0.2);
  height: 1px;
  border: 0;
  margin-bottom: 20px;
  margin-top: 20px;
}

.Input {
  border: 1px solid rgba(0, 0, 0, 0.2);
  border-radius: 3px;
  font-size: 14px;
  margin-bottom: 10px;
  padding: 10px;
  width: 100%;

  &:focus {
    outline-color: #fcc169;
    outline-offset: 2px;
    outline-style: solid;
    outline-width: 2px;
  }
}
