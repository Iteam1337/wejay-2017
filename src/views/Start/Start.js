// @flow

import * as React from 'react'
import * as Wejay from 'typings/wejay.flow'
import * as WejayApi from './__generated__/StartQuery'
import Rooms from './Rooms/RoomsContainer'
import AddRoom from './AddRoom/AddRoomContainer'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import AddUserForm from './AddUserForm'
import type { FormValues } from './AddUserForm'
import * as storage from '../../utils/storage'
import md5 from 'md5'
import styled from 'styled-components'

type Props = {
  data: Wejay.ApolloBase<WejayApi.StartQuery>,
}

type State = {
  hasUser: boolean,
}

const Wrap = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
  height: 100vh;
  justify-content: center;
`

const Content = styled.div`
  max-width: 300px;
`

const Separator = styled.div`
  background-color: rgba(0, 0, 0, 0.2);
  height: 1px;
  border: 0;
  margin-bottom: 20px;
  margin-top: 20px;
`

export class Start extends React.Component<Props, State> {
  state = {
    hasUser: false,
  };

  async componentDidMount () {
    const id = localStorage.getItem('id')

    if (id && !!await storage.getValue(id)) {
      this.hasUser()
    }
  }

  saveUser = async ({ email }: FormValues) => {
    const id = md5(email)

    await storage.put({ id, email })
    localStorage.setItem('id', id)

    this.hasUser()
  };

  hasUser = () => {
    this.setState({
      hasUser: true,
    })
  };

  render () {
    const { data: { error, loading, rooms } } = this.props
    const { hasUser } = this.state

    if (loading) {
      return <div className="Loader">Loading</div>
    }

    if (error) {
      return <div className="Error">{error.message}</div>
    }

    return (
      <Wrap>
        <Content>
          {!hasUser && <AddUserForm saveUser={this.saveUser} />}

          {hasUser && (
            <React.Fragment>
              <Rooms rooms={rooms} />
              <Separator />
              <AddRoom />
            </React.Fragment>
          )}
        </Content>
      </Wrap>
    )
  }
}

export const StartQuery = gql`
  query StartQuery {
    rooms {
      name
    }
  }
`

export default graphql(StartQuery)(Start)
