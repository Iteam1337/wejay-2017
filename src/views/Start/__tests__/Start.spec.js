import React from 'react'
import { shallow } from 'enzyme'
import { Start } from '../Start'

describe('components/Start', () => {
  let component
  let instance

  beforeEach(() => {
    component = shallow(
      <Start
        data={{
          loading: true
        }}
      />
    )

    instance = component.instance()
  })

  it('renders loading message', () => {
    expect(component).toMatchSnapshot()
  })

  it('renders error message', () => {
    component.setProps({
      data: {
        error: {
          message: 'error'
        }
      }
    })

    expect(component).toMatchSnapshot()
  })

  it('renders Start', () => {
    component.setProps({
      data: {
        loading: false,
        error: null,
        rooms: []
      }
    })

    expect(component).toMatchSnapshot()
  })

  it('renders Start if user is set', () => {
    component.setProps({
      data: {
        loading: false,
        error: null,
        rooms: []
      }
    })

    component.setState({
      hasUser: true
    })

    expect(component).toMatchSnapshot()
  })

  describe('#saveEmail', () => {
    it('saves email to localStorage', () => {
      component.setState({
        user: 'cookie@monster.com'
      })

      instance.saveEmail()

      expect(localStorage.setItem).toHaveBeenCalledWith(
        'user',
        'cookie@monster.com'
      )
    })

    it('sets has user to true', () => {
      component.setState({
        user: 'cookie@monster.com'
      })

      instance.saveEmail()

      expect(component.state().hasUser).toBe(true)
    })
  })

  describe('#updateEmail', () => {
    it('sets email on state', () => {
      const event = {
        target: {
          value: 'cookie@monster.com'
        }
      }

      instance.updateEmail(event)

      expect(component.state().user).toEqual('cookie@monster.com')
    })
  })
})
