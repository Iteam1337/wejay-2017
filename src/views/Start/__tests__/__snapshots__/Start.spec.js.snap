// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Start renders Start 1`] = `
<div
  className="Start"
>
  <div
    className="Start__content"
  >
    <div
      className="Start__rooms"
    >
      <div>
        <input
          className="Input"
          onChange={[Function]}
          placeholder="E-mail"
        />
        <button
          className="Rooms__button"
          disabled={true}
          onClick={[Function]}
        >
          Save user
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/Start renders Start if user is set 1`] = `
<div
  className="Start"
>
  <div
    className="Start__content"
  >
    <div
      className="Start__rooms"
    >
      <div>
        <withRouter(Apollo(Rooms))
          rooms={Array []}
        />
        <hr
          className="Start__separator"
        />
        <Apollo(AddRoom) />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/Start renders error message 1`] = `
<div>
  error
</div>
`;

exports[`components/Start renders loading message 1`] = `
<div>
  Loading
</div>
`;
